rclcpp/include/rclcpp/executor.hpp

```diff
@@ -368,21 +360,25 @@ class Executor
     // Put the executor's guard condition in
     guard_condition_handles.guard_conditions[1] = \
       interrupt_guard_condition_->data;
-    // Then fill the SubscriberHandles with ready subscriptions
-    size_t guard_cond_handle_index = start_of_timer_guard_conds;
-    for (auto & timer : timers) {
-      guard_condition_handles.guard_conditions[guard_cond_handle_index] = \
-        timer->guard_condition_->data;
-      guard_cond_handle_index += 1;
-    }
 
     rmw_time_t * wait_timeout = NULL;
     rmw_time_t rmw_timeout;
 
-    if (timeout >= std::chrono::duration<int64_t, T>::zero()) {
+    auto next_timer_duration = get_earliest_timer();
+    /* If the next timer timeout must preempt the requested timeout
+       or if the requested timeout blocks forever, and there exists a valid timer,
+       replace the requested timeout with the next timeout. */
+    if ( (next_timer_duration < timeout || timeout <
+      std::chrono::duration<int64_t,
+      T>::zero()) && next_timer_duration >= std::chrono::nanoseconds::zero())
```

Comment:
If you want to use whitespace, e.g. `if ( (...` to separate things out, please just make them symmetric. So, put a whitespace before the final `)` since there is one proceeding the first `(`.

In general these multiline conditionals are pretty hard to read, and there not much you can do about that. Maybe using some more vertical space would make it easier, maybe like this:

```c++
    if (
      (
        next_timer_duration < timeout ||
        timeout < std::chrono::duration<int64_t, T>::zero()
      ) &&
      next_timer_duration >= std::chrono::nanoseconds::zero()
    ) {
      // ...
    }
```

Honestly my suggestion doesn't look much better, but maybe it will inspire you to come up with a different solution.